kamon-kafka-clients/src/main/scala/kamon/kafka/instrumentation/ConsumerInstrumentation.scala

```diff
@@ -51,33 +49,38 @@ object RecordProcessor {
 
       val consumerSpansForTopic = new mutable.LinkedHashMap[String, Span]()
 
-      records.partitions().forEach(partition => {
+      records.partitions().asScala.foreach(partition => {
         val topic = partition.topic
 
-        records.records(partition).forEach(record => {
+        records.records(partition).asScala.foreach(record => {
           val header = Option(record.headers.lastHeader("kamon-context"))
 
-          val currentContext = header.map(h => Kamon.contextCodec.Binary.decode(ByteBuffer.wrap(h.value))).getOrElse(Context.Empty)
+          val currentContext = header.map{ h =>
+            ContextSerializationHelper.fromByteArray(h.value())
+          }.getOrElse(Context.Empty)
 
           val span = consumerSpansForTopic.getOrElseUpdate(topic, {
-            val spanBuilder = Kamon.buildSpan("poll")
-              .withMetricTag("span.kind", "consumer")
-              .withTag("kafka.partition", partition.partition)
-              .withTag("kafka.topic", topic)
-              .withTag("kafka.offset", record.offset)
-              .withFrom(instant)
-
-            if(Kafka.followStrategy) spanBuilder.asChildOf(currentContext.get(Span.ContextKey))
+            val spanBuilder = Kamon.spanBuilder("poll")
+              .tag("span.kind", "consumer")
+              .tag("kafka.partition", partition.partition)
+              .tag("kafka.topic", topic)
+              .tag("kafka.offset", record.offset)
+
+            // Key could be optional ... see tests
+            Option(record.key()).foreach(k => spanBuilder.tag("kafka.key", record.key().toString))
+
+            if(Kafka.followStrategy) spanBuilder.asChildOf(currentContext.get(Span.Key))
             else {
-              val context = currentContext.get(Span.ContextKey).context()
-              spanBuilder.withTag("trace.related.trace_id", context.traceID.string)
-              spanBuilder.withTag("trace.related.span_id", context.spanID.string)
+              val currentSpan = currentContext.get(Span.Key)
+              spanBuilder
+                .tag("trace.related.trace_id", currentSpan.id.string)
+                .tag("trace.related.span_id", currentSpan.trace.id.string)
             }
             spanBuilder.start()
```

Comment:
Regardless of whether we do delayed spans or not (see the comment attached to the PR), the actual start time for these Spans should be the time at which the "poll" method was called, so we have an idea of how long did the roundtrip to Kafka take.

Reply:
Fixed by passing in the start time from the onEntry() advice